python/modules/TopNNReco.py

```diff
@@ -0,0 +1,475 @@
+        
+    def getGenLevelTopDecays(self,event):
+        genParticles = Collection(event,"GenPart")
+        
+        def findTopIdx(p):
+            motherIdx = p.genPartIdxMother
+            while (motherIdx>=0):
+                if abs(genParticles[motherIdx].pdgId)==6:
+                    return motherIdx
+                motherIdx = genParticles[motherIdx].genPartIdxMother
+            return -1
+            
+        topDict = {}
+        
+        #NB: following will not handle hadronic taus cases
+        for topIdx,genParticle in enumerate(genParticles):
+            if abs(genParticle.pdgId)==6:
+                topDict[topIdx] = {'top': genParticle, 'bquark':[], 'lepton': [], 'neutrino':[], 'quarks': []}
+        
+        for genParticle in genParticles:
+            if abs(genParticle.pdgId)==5 and isLastCopy(genParticle) and fromHardProcess(genParticle):
+                topIdx = findTopIdx(genParticle)
+                if topIdx>=0 and topIdx in topDict.keys():
+                    topDict[topIdx]['bquark'].append(genParticle)
+                    
+            elif abs(genParticle.pdgId)<5 and abs(genParticle.pdgId)>0 and isLastCopy(genParticle) and fromHardProcess(genParticle):
+                topIdx = findTopIdx(genParticle)
+                if topIdx>=0 and topIdx in topDict.keys():
+                    topDict[topIdx]['quarks'].append(genParticle)
+                       
+            elif abs(genParticle.pdgId) in [11,13] and isPrompt(genParticle) and isLastCopy(genParticle):
+                topIdx = findTopIdx(genParticle)
+                if topIdx>=0 and topIdx in topDict.keys():
+                    topDict[topIdx]['lepton'].append(genParticle)
+                    
+            elif abs(genParticle.pdgId) in [12,14] and isPrompt(genParticle) and isLastCopy(genParticle):
+                topIdx = findTopIdx(genParticle)
+                if topIdx>=0 and topIdx in topDict.keys():
+                    topDict[topIdx]['neutrino'].append(genParticle)
+            
+        hadronicTops = []
+        leptonicTops = []
+        otherTops = []
+            
+        for idx in topDict.keys():
+            top = topDict[idx]
+            if len(top['quarks'])==2 and len(top['lepton'])==0 and len(top['neutrino'])==0 and len(top['bquark'])==1:
+                top['quarks'] = sorted(top['quarks'],key=lambda x: x.pt, reverse=True) #sort by pT like jets
+                hadronicTops.append(top)
+                
+            elif len(top['quarks'])==0 and len(top['lepton'])==1 and len(top['neutrino'])==1 and len(top['bquark'])==1:
+                leptonicTops.append(top)
+            else:
+                # rare more complex decays (e.g. involving taus or FSR)
+                otherTops.append(top)
+                
+        return {"hadronic": hadronicTops, "leptonic": leptonicTops, "other": otherTops}
+
+    def makeLeptonJetCombinations(self, leptons, jets):
+        combinations = []
+        for lepton in leptons:
```

Comment:
Consider using `from itertools import permutations` to obtain all jet lepton permutations:
```
perm = permutations(leptons,jets)
```

Reply:
better to allow for additional filtering in the loop